lib/signs_ui/signs/expiration.ex

```diff
@@ -0,0 +1,82 @@
+defmodule SignsUI.Signs.Expiration do
+  @moduledoc """
+  Handle cleaning up settings in the configuration that have since expired
+  """
+  use GenServer
+  alias SignsUI.Signs.Sign
+
+  @type state :: %{
+          time_fetcher: (() -> DateTime.t()),
+          loop_ms: integer(),
+          sign_state_server: GenServer.server()
+        }
+
+  @spec start_link(Keyword.t()) :: GenServer.on_start()
+  def start_link(opts \\ []) do
+    name = Keyword.get(opts, :name, __MODULE__)
+
+    GenServer.start_link(__MODULE__, opts, name: name)
+  end
+
+  @spec init(Keyword.t()) :: {:ok, state()}
+  def init(opts) do
+    time_fetcher = Keyword.get(opts, :time_fetcher, fn -> DateTime.utc_now() end)
+    loop_ms = Keyword.get(opts, :loop_ms, 5_000)
+    sign_state_server = Keyword.get(opts, :sign_state_server, SignsUI.Signs.State)
+    schedule_loop(self(), loop_ms)
+    {:ok, %{time_fetcher: time_fetcher, loop_ms: loop_ms, sign_state_server: sign_state_server}}
+  end
+
+  @spec handle_info(:process_expired, state()) :: state()
+  def handle_info(:process_expired, state) do
+    expire_signs(state.time_fetcher, state.sign_state_server)
+
+    schedule_loop(self(), state.loop_ms)
+
+    {:noreply, state}
+  end
+
+  @spec expire_signs((() -> DateTime.t()), GenServer.server()) :: :ok
+  def expire_signs(time_fetcher, sign_state_server) do
```

Comment:
Actually, since this is the real "meat" of the server, what do you think about extracting out the core of it into a pure function, and move the global state interactions to the `handle_info` loop? That should make it easier to test a few more cases. That's the convention we try to shoot for in most of our code. (e.g. in `RTR`, the `RTR.Vehicles.State` module actually stores and retrieves the vehicles, and then the `RTR.Vehicles.Vehicles` module has a bunch of pure functions that take vehicles as an input and return different ones as the output.)

This function could receive the `get_all()` signs as an argument and then return its calculated `expired_signs`. The `get_all()` and the `if != {}` stuff would be in the `handle_info(:process_expired)` part.

In the `expiration_test.exs` file, you can then very easily pass in a few variety of signs in a map, some with `nil` and others with DateTimes in the past or future, etc, and assert that the correctly changed signs come out of the function.

The more end-to-end sort of test that you have now in `expiration_test.exs` could then be updated to be a test of `handle_info(:process_expired, ..)` instead, with minimal changes.

Reply:
That sounds like an excellent idea for after I eat lunch. 🌯